### PR TITLE
Split up the representations of the C and C++ compilers

### PR DESCRIPTION
**Pull request by dcbaker**
This is part of the work I've been doing to add support for ICL (the Intel C/C++ compiler for windows) to meson. Along the way I've pushed our current representation of the Visual Studio like compilers past their limits, and run into a lot of problems causes by multiple inheritance. One of the biggest headaches is that the C++ compilers inherit the C compilers, and then override what they don't want. Unfortunately this leads to a number of cases of diamond inheritance, such as `CPPCompiler(CCompiler)`, `VisualStudioCCompiler(CCompiler)`, and `VisualStudioCPPCompiler(VisualStudioCCompiler, CPPCompiler)`. This gives us two different places to end up with methods from the CCompiler, and once I started trying to break ClangCl and VisualStudio up it just ends up being madness.

I've instead opted for a more mixin like inheritance scheme, base classes generally don't inherit from more than one parent class, and all multiple inheritance is done in the final class. A mixin provides code sharing for CCompler and CPPCompiler, and a set of VisualStudioLikeCompiler mixins provide specific implementation details for ClangCl and VisualStudio with more specifics for C vs CPP. There is probably more refactoring that could be done here to break some of the GCCisms out of the base C and CPPCompiler classes, but this patch is already 3000 lines of churn and I didn't want to do too much at once.

### CODE REVIEW

**Comment by Ericson2314** [comment]
This is very good! I hope it will be possible to rebase your dynamic linker PR after these two :)

**Comment by dcbaker** [comment]
Fixed ObjC and ObjC++ by making them inherit the CLikeCompiler class instead of the CCompiler and CPPCompiler respectively.

**Comment by Ericson2314** [comment]
@dcbaker Fortran missing things like `concatenate_string_literals` makes me think might need a `WithCPreProcessorCompiler` mixin? That could perhaps help with the CPPFLAGS issues @nirbheek was trying to solve before, too.

**Comment by dcbaker** [comment]
@Ericson2314 there's a lot of layering violations going on, for the moment I've just made Fortran inherit the CLikeCompiler class because the way Fortran was implemented was really gross. There's some stuff that needs to make it's way into the GnuLikeCompiler class, and probably a couple of mixins that need to be split out (the CPreProcessor is one, I think maybe the RPATH handling stuff, Maybe a couple of others.)

**Comment by Ericson2314** [comment]
@dcbaker Sounds good. If it works, it works, and delaying that stuff to a futuredoesn't keep this from being a strict improvement over the status quo.

**Comment by dcbaker** [comment]
@nirbheek, @jpakkane any comments?

File: mesonbuild/compilers/clike.py [hunk 1]
```diff
@@ -0,0 +1,1179 @@
...
+        patterns = self.get_library_naming(env, libtype)
+        # try to detect if we are 64-bit or 32-bit. If we can't
+        # detect, we will just skip path validity checks done in
+        # get_library_dirs() call
+        try:
+            if self.output_is_64bit(env):
+                elf_class = 2
+            else:
+                elf_class = 1
+        except (MesonException, KeyError): # TODO evaluate if catching KeyError is wanted here
```
Comment (ignatenkobrain) [inline mesonbuild/compilers/clike.py]: __[Flake8]__ [F821] undefined name 'MesonException' ([link](https://sider.review/gh/repos/19784232/pulls/5339#issue-4029462))

<sub>posted by Sider</sub>

**Comment by ignatenkobrain** [comment]
Flake8 detected 20 issues on 00e7e3149c16c70831551d61d3e8d27cf5dd719c
Visit https://sider.review/gh/repos/19784232/pulls/5339 to review the issues.

<sub>posted by Sider</sub>

**Comment by dcbaker** [comment]
@jpakkane, @nirbheek I really don't want to rebase this again, anyone object to merging this?

**Comment by jpakkane** [comment]
Good stuff, thanks.

**Pull request merged**